### PR TITLE
janet: 1.16.1 -> 1.20.0

### PR DESCRIPTION
**Pull request by andrewchambers**
###### Motivation for this change

New janet release.

### CODE REVIEW

File: pkgs/development/interpreters/janet/default.nix [hunk 1]
```diff
@@ -1,28 +1,15 @@
...
-  # we don't have /usr/bin/env in the sandbox, so substitute for a proper,
-  # absolute path to janet
-  postPatch = ''
-    substituteInPlace jpm \
-      --replace '/usr/bin/env janet' $out/bin/janet \
-      --replace /usr/local/lib/janet $out/lib \
-      --replace /usr/local           $out
-
-    substituteInPlace janet.1 \
-      --replace /usr/local/lib/janet $out/lib
```
Comment (taku0) [inline pkgs/development/interpreters/janet/default.nix:23]: Is it safe to delete this?

**Comment by taku0** [comment]
It works fine:
- `(+ 1 2 3)`
- `(print "Hello, World!")`
- `(os/exit)`

File: pkgs/development/interpreters/janet/default.nix (hunk 1 above)
Comment (andrewchambers) [reply pkgs/development/interpreters/janet/default.nix:23]: It was updating a man page, but the update doesn't really affect anything useful imo.

Comment (taku0) [reply pkgs/development/interpreters/janet/default.nix:23]: What is the actual default value of `syspath`? `/usr/local/lib/janet` or `$out/lib`?

Comment (andrewchambers) [reply pkgs/development/interpreters/janet/default.nix:23]: It depends what you mean by default? If we change it is it really the default now?

Comment (taku0) [reply pkgs/development/interpreters/janet/default.nix:23]: I mean `(dyn :syspath)` and it is `/nix/store/491jv1pqvlbkg04wjf933k6dka9snqrg-janet-1.19.2/lib/janet`, so the sentence should be like following:

```diff
- defaults to /usr/local/lib/janet on Linux/Posix, and C:/Janet/Library on Windows.
+ defaults to /nix/store/491jv1pqvlbkg04wjf933k6dka9snqrg-janet-1.19.2/lib/janet.
```

Comment (andrewchambers) [reply pkgs/development/interpreters/janet/default.nix:23]: I know, but if upstream wants that they should patch their own man page at install time, I don't know why you want to rewrite upstreams man pages. You could submit a patch to them if you want that.

Comment (taku0) [reply pkgs/development/interpreters/janet/default.nix:23]: Ah, the sentence can be read as “the default value of the default value is ...”, then “/usr/local/lib/janet on Linux/Posix” makes sense even on NixOS.

Comment (andrewchambers) [reply pkgs/development/interpreters/janet/default.nix:23]: That was my interpretation, I guess I was not so clear rereading.

**Comment by taku0** [comment]
The test `test/suite0009.janet` fails on macOS.

```
[0/1] Running all tests.
 1/11 test/suite0000.janet         OK              0.03s
 2/11 test/suite0002.janet         OK              0.02s
 3/11 test/suite0001.janet         OK              0.04s
 4/11 test/suite0003.janet         OK              0.03s
 5/11 test/suite0004.janet         OK              0.03s
 6/11 test/suite0005.janet         OK              0.03s
 7/11 test/suite0006.janet         OK              0.03s
 8/11 test/suite0008.janet         OK              0.03s
 9/11 test/suite0007.janet         OK              0.05s
10/11 test/suite0010.janet         OK              0.03s
11/11 test/suite0009.janet         TIMEOUT        30.02s   killed by signal 15 SIGTERM
>>> MALLOC_PERTURB_=216 /private/tmp/nix-build-janet-1.19.2.drv-0/source/build/janet /private/tmp/nix-build-janet-1.19.2.drv-0/source/build/../test/suite0009.janet
 ✀  
stderr:
Starting suite 9...



Summary of Failures:

11/11 test/suite0009.janet TIMEOUT        30.02s   killed by signal 15 SIGTERM


Ok:                 10  
Expected Fail:      0   
Fail:               0   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            1   
```

**Comment by andrewchambers** [comment]
I wonder if this is caused by the Nix sandboxing? I am not familiar with mac.

**Comment by taku0** [comment]
`1.16.1` builds fine.

The stuck test is this: https://github.com/janet-lang/janet/blob/1c937ad960d3319f7dad6a534046e2f7f4aaca6c/test/suite0009.janet#L178-L184

The official binary distribution runs the test without errors.

**Comment by andrewchambers** [comment]
AFAIK these are new tests that weren't in 1.16.1, so that isn't too relevant. I personally don't mind disabling mac support until someone with a mac can debug it.

**Comment by taku0** [comment]
`net/write` don't return if `net/read` is being called in another fiber.  If we insert `(ev/sleep 1)` before `net/read`, it works.  The official binary don't suffer from this; it works even if `(ev/sleep 1)` is inserted before `net/write`.

1.16.1 also have this problem.  I propose disabling that test and merging this PR anyway.

**Comment by andrewchambers** [comment]
@taku0 does the test need to be disabled if the package is marked broken? I have added the broken change.

**Comment by cideM** [comment]
This patch

```
diff --git a/test/suite0009.janet b/test/suite0009.janet
index 6095bc60..25360d60 100644
--- a/test/suite0009.janet
+++ b/test/suite0009.janet
@@ -174,15 +174,6 @@
   (defer (:close stream)
     (check-matching-names stream)))
 
-# Test localname and peername
-(repeat 20
-  (with [s (net/server "127.0.0.1" "8000" names-handler)]
-    (defn test-names []
-      (with [conn (net/connect "127.0.0.1" "8000")]
-        (check-matching-names conn)))
-    (repeat 20 (test-names)))
-  (gccollect))
-
 # Create pipe
 
 (var pipe-counter 0)
```

fixes the build on my M1 machine.

Also this update will remove `jpm`. Are we ok with that? It would have to be bundled separately then.

**Comment by andrewchambers** [comment]
Disabling tests might just be hiding the problem - it would be better to work out the cause and fix it upstream.

**Comment by peterhoeg** [comment]
> Also this update will remove `jpm`. Are we ok with that? It would have to be bundled separately then.

We absolutely should do jpm as a separate package with this.

**Comment by andrewchambers** [comment]
Makes sense, I will add the patch when we are on macos.

**Comment by andrewchambers** [comment]
@cideM I added your patch and added you as a co-author, I hope that is ok.

File: pkgs/development/interpreters/janet/default.nix [hunk 2]
```diff
@@ -1,39 +1,29 @@
...
-  '';
-
   nativeBuildInputs = [ meson ninja ];
 
   mesonFlags = [ "-Dgit_hash=release" ];
 
   doCheck = true;
 
+  patches = lib.optionals stdenv.isDarwin ./darwin-remove-net-test.patch;
```
Comment (taku0) [inline pkgs/development/interpreters/janet/default.nix:19]: Please refer this pull request in a comment.

File: pkgs/development/interpreters/janet/default.nix [hunk 3]
```diff
@@ -1,39 +1,29 @@
...
+
   meta = with lib; {
     description = "Janet programming language";
     homepage = "https://janet-lang.org/";
     license = licenses.mit;
     maintainers = with maintainers; [ andrewchambers peterhoeg ];
     platforms = platforms.all;
+    broken = stdenv.isDarwin;
```
Comment (taku0) [inline pkgs/development/interpreters/janet/default.nix:27]: Also here, add a comment referring to this pull request.

Comment (andrewchambers) [reply pkgs/development/interpreters/janet/default.nix:30]: done.

**Review by taku0 — APPROVED** [review]

**Comment by cideM** [comment]
Tested compilation on my M1 machine with `NIXPKGS_ALLOW_BROKEN=1 nix-build -A janet .` and `(+ 1 2 3)` and works fine 👍🏻

**Comment by peterhoeg** [comment]
Anyone feels like taking a stab at jpm so we can get this merged as well?

**Comment by andrewchambers** [comment]
@peterhoeg do you mean this shouldn't be merged until jpm is packaged? I was hoping they could be separate.

**Comment by cideM** [comment]
I still have zero knowledge of Janet but it seems to me that if this is merged we end up with a Janet without any package manager in Nixpkgs, right?

**Comment by andrewchambers** [comment]
@cideM Your choice is an outdated janet or one without jpm unless you want to do the jpm package yourself.

**Comment by peterhoeg** [comment]
> but it seems to me that if this is merged we end up with a Janet without any package manager in Nixpkgs, right?

That was my point as well. At least things *work* right now (albeit on an older version), whereas no jpm is most definitely a regression.

**Comment by andrewchambers** [comment]
I would argue is not a regression, but it is a breaking change - jpm is no longer included with janet, it is a separate program. I agree a jpm package would be nice, but it is going to be a breaking change regardless since people will need to update their nix configs.

**Comment by peterhoeg** [comment]
> I don't know when that will be.
Right about now! ;-)

**Comment by peterhoeg** [comment]
@cideM , I just ripped out aarch64 as it wasn't evaluating there - it's just temporary.

**Comment by andrewchambers** [comment]
Really appreciated! thank you.

**Comment by peterhoeg** [comment]
I need to squash these commits but it requires me to force push to your branch @andrewchambers - ok with that?

**Comment by andrewchambers** [comment]
Sure I don't mind.

**Pull request merged**